### PR TITLE
Generate button letters A, B, E, H, P, R, S and new Battle Only button

### PR DESCRIPTION
**Pull request by zenseii**
Close https://github.com/ihhub/fheroes2/issues/5234

With letter spacing fixed we can
close https://github.com/ihhub/fheroes2/issues/5858

The 'A' is different from the OG since the automatic shadowing etc is different and I had to add a pair of pixels to fill some gaps.
image

~~In the future it should be made possible to have the letter spacing be 2px smaller (i.e. 1px on each side of a letter) so that the buttons matche the OG main menu buttons:~~ This has been achieved in this PR.

image

image

Aligned and tighter Difficulty button:
image

### CODE REVIEW

**Comment by zenseii** [comment]
Current issues with translation is that if the translated string contains either `\n` or a white space then they will count as not valid characters and the English text will be used.

**Comment by ihhub** [comment]
Hi @zenseii , to fix 2 pixel space we need to apply a shift for them. Try to add this code at the end of `generateBaseButtonFont()` function:
```cpp
for ( Sprite & letter : released ) {
    letter.setPosition( -2, 0 );
}

for ( Sprite & letter : pressed ) {
    letter.setPosition( -2, 0 );
}
```
I think it should help to solve your concern.

**Review by ihhub — APPROVED** [review]

File: src/fheroes2/agg/agg_image.cpp [hunk 1]
```diff
@@ -355,15 +355,19 @@ namespace fheroes2
...
                 _icnVsSprite[id].resize( 2 );
                 for ( int32_t i = 0; i < static_cast<int32_t>( _icnVsSprite[id].size() ); ++i ) {
                     Sprite & out = _icnVsSprite[id][i];
                     out = GetICN( ICN::BTNCOM, i );
-                    // Clean button
+                    // clean button.
                     Fill( out, 13, 11, 113, 31, getButtonFillingColor( i == 0 ) );
                 }
+
+                renderTextOnButton( _icnVsSprite[id][0], _icnVsSprite[id][1], gettext_noop( "BATTLE\nONLY" ), { 14, 29 }, { 13, 30 }, 110, fheroes2::FontColor::WHITE );
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:367]: Hi @zenseii , the problem with not centering is in slightly incorrect values here. We need to take into account that the area for the text is this:
image

To make the button to be vertically and horizontally aligned we can pass fheroes2::Rect values. For this case it should be:
```cpp
{ 11, 4, 117, 47 }, { 10, 5, 117, 47 }
```
However, size of a button doesn't change between states so we pass offsets and size:
```cpp
{ 11, 4 }, { 10, 5 }, { 117, 47 }
```
Then `renderTextOnButton` function should be:
```cpp
    void renderTextOnButton( fheroes2::Image & releasedState, fheroes2::Image & pressedState, const char * text, const fheroes2::Point & releasedTextOffset,
                             const fheroes2::Point & pressedTextOffset, const fheroes2::Size buttonSize, const fheroes2::FontColor fontColor )
    {
        const fheroes2::FontType releasedFont{ fheroes2::FontSize::BUTTON_RELEASED, fontColor };
        const fheroes2::FontType pressedFont{ fheroes2::FontSize::BUTTON_PRESSED, fontColor };

        const char * textSupported = getSupportedText( text, releasedFont );

        fheroes2::Text releasedText( textSupported, releasedFont );
        fheroes2::Text pressedText( textSupported, pressedFont );

        const fheroes2::Size releaseTextSize( releasedText.width( buttonSize.width ), releasedText.height( buttonSize.width ) );
        const fheroes2::Size pressedTextSize( pressedText.width( buttonSize.width ), pressedText.height( buttonSize.width ) );

        releasedText.draw( releasedTextOffset.x + ( buttonSize.width - releaseTextSize.width ) / 2, releasedTextOffset.y + ( buttonSize.height - releaseTextSize.height ) / 2, buttonSize.width, releasedState );
        pressedText.draw( pressedTextOffset.x + ( buttonSize.width - pressedTextSize.width ) / 2, pressedTextOffset.y + ( buttonSize.height - pressedText.height / 2 ), buttonSize.width, pressedState );
    }
```

The only drawback is possible 2 pixels shift of overall text to the left. What do you think?

Comment (zenseii) [reply src/fheroes2/agg/agg_image.cpp:367]: BTW why is the rectangle 47px tall and not 48px in this case? This is just so I know how to proceed with the others.
image

Comment (ihhub) [reply src/fheroes2/agg/agg_image.cpp:367]: Actually, because pressed state has 47 pixels height. My apologies for not explaining this.

image

Comment (zenseii) [reply src/fheroes2/agg/agg_image.cpp:367]: I should've known these sizes by heart by now...

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@zenseii , huge thanks for this useful improvement!

**Pull request merged**